curly: "multi-or-nest" option.

I've long been a fan of the (seemingly unpopular) curly style that allows omitting the braces for a single statement.  However, I like to enforce braces if inside another block, for example:

```javascript
// Acceptable
for (let i = 0; i < l; i++) {
    if (a[i])
        fn();
}

if (cond) {
    while (cond2) {
        for (cond3 in foo)
            bar();
    }
}
baz();

// Should be a warning
for (let i = 0; i < l; i++)
    if (a[i])
        fn();

if (cond)
    while (cond2)
        for (cond3 in foo)
            bar();
baz();
```

Generally, I find braces to be unnecessary clutter, but I like to prevent a large de-indent when continuing on after nested blocks.  Currently, I can't enforce this style with eslint.  The first form gives a warning regardless of whether the `curly` is enabled or disabled.

Any chance this could be implemented?

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/8403809-curly-multi-or-nest-option?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

This is a sample of some code I have:

```javascript
for (let i = 0, m = modules.length; i < m; i++) {
    if (modules[i].domains.some(matchHost))
        return makeAPIRequest(modules[i], urlParts).then(resolve, reject);
}
```

This is the error eslint gives with `curly: [2, "multi"]`:

```
   41:12  error  Unnecessary { after 'for' condition    curly
```

--- a/docs/rules/curly.md
+++ b/docs/rules/curly.md
@@ -138,6 +138,67 @@ while (true) {
 }
 ```
 
+You can use another configuration that forces brace-less `if`, `else if`, `else`, `for`, `while`, or `do` if their body contains only one single-line statement. And forces braces in all other cases.
+
+```json
+curly: [2, "multi-or-nest"]
+```
+
+With this configuration, the rule will warn for these patterns:
+
+```js
+if (foo)
+    return {
+        bar: baz,
+        qux: foo
+    };
+
+while (true)
+  if(foo)
+      doSomething();
+  else
+      doSomethingElse();
+
+if (foo) {
+    return;
+}
+
+while (true) {
+    doSomething();
+}
+
+for (var i = 0; foo; i++) {
+    doSomething();
+}
+```
+
+It will not warn for these patterns:
+
+```js
+if (foo) {
+    return {
+        bar: baz,
+        qux: foo
+    };
+}
+
+while (true) {
+  if(foo)
+      doSomething();
+  else
+      doSomethingElse();
+}
+
+if (foo)
+    return;
+
+while (true)
+    doSomething();
+
+for (var i = 0; foo; i++)
+    doSomething();
+```
+
 The default configuration is:
 
 ```json

--- a/lib/rules/curly.js
+++ b/lib/rules/curly.js
@@ -1,6 +1,7 @@
 /**
  * @fileoverview Rule to flag statements without curly braces
  * @author Nicholas C. Zakas
+ * @copyright 2015 Ivan Nikulin. All rights reserved.
  */
 "use strict";
 
@@ -12,6 +13,7 @@ module.exports = function(context) {
 
     var multiOnly = (context.options[0] === "multi");
     var multiLine = (context.options[0] === "multi-line");
+    var multiOrNest = (context.options[0] === "multi-or-nest");
 
     //--------------------------------------------------------------------------
     // Helpers
@@ -30,6 +32,52 @@ module.exports = function(context) {
     }
 
     /**
+     * Determines if a given node is a one-liner.
+     * @param {ASTNode} node The node to check.
+     * @returns {boolean} True if the node is a one-liner.
+     * @private
+     */
+    function isOneLiner(node) {
+        var first = context.getFirstToken(node),
+            last = context.getLastToken(node);
+
+        return first.loc.start.line === last.loc.end.line;
+    }
+
+    /**
+     * Reports "Expected { after ..." error
+     * @param {ASTNode} node The node to report.
+     * @param {string} name The name to report.
+     * @param {string} suffix Additional string to add to the end of a report.
+     * @returns {void}
+     * @private
+     */
+    function reportExpectedBraceError(node, name, suffix) {
+        context.report(node, "Expected { after '{{name}}'{{suffix}}.",
+            {
+                name: name,
+                suffix: (suffix ? " " + suffix : "")
+            });
+    }
+
+    /**
+     * Reports "Unnecessary { after ..." error
+     * @param {ASTNode} node The node to report.
+     * @param {string} name The name to report.
+     * @param {string} suffix Additional string to add to the end of a report.
+     * @returns {void}
+     * @private
+     */
+    function reportUnnecessaryBraceError(node, name, suffix) {
+        context.report(node, "Unnecessary { after '{{name}}'{{suffix}}.",
+            {
+                name: name,
+                suffix: (suffix ? " " + suffix : "")
+            }
+        );
+    }
+
+    /**
      * Checks the body of a node to see if it's a block statement. Depending on
      * the rule options, reports the appropriate problems.
      * @param {ASTNode} node The node to report if there's a problem.
@@ -43,30 +91,21 @@ module.exports = function(context) {
 
         if (multiOnly) {
             if (hasBlock && body.body.length === 1) {
-                context.report(node, "Unnecessary { after '{{name}}'{{suffix}}.",
-                    {
-                        name: name,
-                        suffix: (suffix ? " " + suffix : "")
-                    }
-                );
+                reportUnnecessaryBraceError(node, name, suffix);
             }
         } else if (multiLine) {
             if (!hasBlock && !isCollapsedOneLiner(body)) {
-                context.report(node, "Expected { after '{{name}}'{{suffix}}.",
-                    {
-                        name: name,
-                        suffix: (suffix ? " " + suffix : "")
-                    }
-                );
+                reportExpectedBraceError(node, name, suffix);
+            }
+        } else if (multiOrNest) {
+            if (hasBlock && body.body.length === 1 && isOneLiner(body.body[0])) {
+                reportUnnecessaryBraceError(node, name, suffix);
+            } else if (!hasBlock && !isOneLiner(body)) {
+                reportExpectedBraceError(node, name, suffix);
             }
         } else {
             if (!hasBlock) {
-                context.report(node, "Expected { after '{{name}}'{{suffix}}.",
-                    {
-                        name: name,
-                        suffix: (suffix ? " " + suffix : "")
-                    }
-                );
+                reportExpectedBraceError(node, name, suffix);
             }
         }
     }
@@ -104,6 +143,6 @@ module.exports = function(context) {
 
 module.exports.schema = [
     {
-        "enum": ["all", "multi", "multi-line"]
+        "enum": ["all", "multi", "multi-line", "multi-or-nest"]
     }
 ];

--- a/tests/lib/rules/curly.js
+++ b/tests/lib/rules/curly.js
@@ -67,7 +67,32 @@ ruleTester.run("curly", rule, {
         {
             code: "do bar() \n while (foo)",
             options: ["multi-line"]
+        },
+        {
+            code: "if (foo) { \n quz = { \n bar: baz, \n qux: foo \n }; \n }",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "while (true) { \n if (foo) \n doSomething(); \n else \n doSomethingElse(); \n }",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "if (foo) \n quz = true;",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "while (true) \n doSomething();",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "for (var i = 0; foo; i++) \n doSomething();",
+            options: ["multi-or-nest"]
+        },
+        {
+            code: "if (foo) { \n if(bar) \n doSomething(); \n } else \n doSomethingElse();",
+            options: ["multi-or-nest"]
         }
+
     ],
     invalid: [
         {
@@ -212,6 +237,56 @@ ruleTester.run("curly", rule, {
                 {
                     message: "Expected { after 'do'.",
                     type: "DoWhileStatement"
+                }
+            ]
+        },
+        {
+            code: "if (foo) \n quz = { \n bar: baz, \n qux: foo \n };",
+            options: ["multi-or-nest"],
+            errors: [
+                {
+                    message: "Expected { after 'if' condition.",
+                    type: "IfStatement"
+                }
+            ]
+        },
+        {
+            code: "while (true) \n if (foo) \n doSomething(); \n else \n doSomethingElse(); \n",
+            options: ["multi-or-nest"],
+            errors: [
+                {
+                    message: "Expected { after 'while' condition.",
+                    type: "WhileStatement"
+                }
+            ]
+        },
+        {
+            code: "if (foo) { \n quz = true; \n }",
+            options: ["multi-or-nest"],
+            errors: [
+                {
+                    message: "Unnecessary { after 'if' condition.",
+                    type: "IfStatement"
+                }
+            ]
+        },
+        {
+            code: "while (true) { \n doSomething(); \n }",
+            options: ["multi-or-nest"],
+            errors: [
+                {
+                    message: "Unnecessary { after 'while' condition.",
+                    type: "WhileStatement"
+                }
+            ]
+        },
+        {
+            code: "for (var i = 0; foo; i++) { \n doSomething(); \n }",
+            options: ["multi-or-nest"],
+            errors: [
+                {
+                    message: "Unnecessary { after 'for' condition.",
+                    type: "ForStatement"
                 }
             ]
         }

